subsys/power/device_pm.c

```diff
@@ -0,0 +1,226 @@
+/*
+ * Copyright (c) 2018 Intel Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr.h>
+#include <kernel.h>
+#include <init.h>
+#include <string.h>
+#include <soc.h>
+#include <device.h>
+#include "policy/pm_policy.h"
+
+#define LOG_LEVEL CONFIG_PM_LOG_LEVEL /* From power module Kconfig */
+#include <logging/log.h>
+LOG_MODULE_DECLARE(power);
+
+/* Async device PM request status */
+#define DEVICE_PM_REQ_STAT_INIT		0
+#define DEVICE_PM_REQ_STAT_CANCEL	1
+#define DEVICE_PM_REQ_STAT_DONE		2
+#define DEVICE_PM_REQ_STAT_ERROR	3
+
+/* Device PM request type */
+#define DEVICE_PM_ASYNC			(1 << 0)
+
+/* device PM worker queue thread info */
+K_THREAD_STACK_DEFINE(devpm_workq_stack, CONFIG_DEVPM_WORKQUEUE_STACK_SIZE);
+struct k_work_q devpm_workq;
+
+/*
+ * Device PM worker to handle the Async PM requests
+ * from device_pm_get/device_pm_put API's. The worker
+ * will call the device_set_power_state() to put the
+ * device in desired power state and notify the caller
+ * through k_alert mechanism.
+ */
+static void pm_work_handler(struct k_work *work)
+{
+	struct device_pm *pm = CONTAINER_OF(work,
+					struct device_pm, dwork);
+	struct device *dev = device_get_binding(pm->name);
+	int ret;
+
+	k_sem_take(&dev->pm->lock, K_FOREVER);
+
+	/* Incase of suspend request, make sure the device is not busy */
+	if (device_busy_check(dev) && (dev->pm->request
+				== DEVICE_PM_SUSPEND_STATE)) {
+		LOG_ERR("dev_pm: device busy!!\n");
+		dev->pm->status =  DEVICE_PM_REQ_STAT_ERROR;
+		goto out;
+	}
+
+	ret = device_set_power_state(dev, dev->pm->request);
+	if (ret) {
+		LOG_ERR("Set power state(%d) error(%d)\n",
+					dev->pm->request, ret);
+		dev->pm->status = DEVICE_PM_REQ_STAT_ERROR;
+	} else {
+		dev->pm->status = DEVICE_PM_REQ_STAT_DONE;
+	}
+out:
+	/* Notify the driver through alert */
+	if (dev->pm->alert.handler) {
+		k_alert_send(&dev->pm->alert);
+	}
+	k_sem_give(&dev->pm->lock);
+}
+
+/*
+ * Set the device PM state to target PM state based on
+ * the current state. If the PM request is of Async type
+ * then a work item will be submitted to PM worker queue
+ * else it will call device_set_power_state() to set the
+ * device PM state to target state.
+ */
+static int device_pm_set_state(struct device *dev,
+			       u32_t target_state, u32_t pm_flags)
+{
+	int ret;
+	u32_t current_state;
+
+	/* Get current device power state */
+	ret = device_get_power_state(dev, &current_state);
+	if (ret) {
+		LOG_ERR("Get power state error(%d)\n", ret);
+		goto out;
+	}
+
+	/* Return if current state is same as target state */
+	if (current_state == target_state) {
```

Comment:
Possible bug. Scenario: 
Async request to suspend followed by sync to wake-up.
1. Asynchronous request is queue in work queue, but not yet executed.
2. Sync request exists as the device is in target state.
3. Asynchronous request is executed, putting device in suspend state while it is used.

Reply:
Ok, I will have to move the worker cancel logic above this block.